p2p/src/swarm/peerdb.rs

```diff
@@ -122,25 +155,64 @@ impl<T: NetworkingService> PeerDb<T> {
         false // TODO: implement
     }
 
-    /// Discover new peer addresses
-    pub fn discover_peers(&mut self, peers: &[types::AddrInfo<T>]) {
-        log::info!("discovered {} new peers", peers.len());
+    /// Check if the peers is part of our active swarm
+    pub fn is_active_peer(&self, peer_id: &T::PeerId) -> bool {
+        std::matches!(self.peers.get(peer_id), Some(Peer::Active(_)))
+    }
 
-        for info in peers.iter() {
-            // TODO: update peer stats
-            if self.peers.contains_key(&info.id) {
-                continue;
-            }
+    /// Get socket address of the next best peer (TODO: in terms of peer score)
+    // TODO: rewrite all of this
+    pub fn best_peer_addr(&mut self) -> crate::Result<Option<T::Address>> {
+        // TODO: improve peer selection
+        let peer_id = match self.available.iter().next() {
+            Some(peer_id) => *peer_id,
+            None => return Ok(None),
+        };
 
-            match self.discovered.entry(info.id).or_insert_with(|| PeerAddrInfo::Raw {
-                ip4: HashSet::new(),
-                ip6: HashSet::new(),
-            }) {
-                PeerAddrInfo::Raw { ip4, ip6 } => {
-                    log::trace!("discovered ipv4 {:#?}, ipv6 {:#?}", ip4, ip6);
+        match self.peers.get_mut(&peer_id) {
+            Some(Peer::Idle(_info) | Peer::Active(_info)) => {
+                // TODO: implement
+                Ok(None)
+            }
+            Some(Peer::Banned(_info)) => {
+                // TODO: implement
+                Ok(None)
+            }
+            Some(Peer::Discovered(addr_info)) => addr_info.pop_front().map_or(Ok(None), |addr| {
+                self.available.remove(&peer_id);
+                self.pending.insert(addr.clone(), peer_id);
+                Ok(Some(addr))
+            }),
+            None => Err(P2pError::DatabaseFailure),
+        }
+    }
 
-                    ip4.extend(info.ip4.clone());
-                    ip6.extend(info.ip6.clone());
+    /// Discover new peer addresses
+    pub fn discover_peers(&mut self, peers: &[types::AddrInfo<T>]) {
```

Comment:
I think this function can be broken down to become more meaningful. Usually functions that take lists and loop over their whole operation can be reduced to functions that take individual elements, which then you use with a map. As I see from the implementation, there's no relation between any of the peers. Maybe it's better to make this function `fn discover_peer(&mut self, peer: types::AddrInfo<T>) -> Result<()>`. Then, if you wish, you can use a helper function, though I'd prefer to see the caller just use a map or for_each with all the peers.